timemachine/cpp/src/device_buffer.hpp

```diff
@@ -0,0 +1,25 @@
+#pragma once
+#include <cstdlib>
+
+namespace timemachine {
+
+template <typename T> class DeviceBuffer {
+
+public:
+    DeviceBuffer(const std::size_t length);
+
+    ~DeviceBuffer();
+
+    const size_t size;
+
+    void ensure_allocated();
+
+    T *data();
```

Comment:
i'd also make a const version of this for read-only access

`const T *data();`